x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java

```diff
@@ -93,29 +102,39 @@
      * Otherwise, <em>N</em> will be added to the <em>local</em> index list.
      */
 
-    ResolvedIndices resolve(TransportRequest request, Metadata metadata, List<String> authorizedIndices) {
+    void resolve(TransportRequest request, Metadata metadata, AsyncSupplier<List<String>> authorizedIndicesSupplier,
+                 ActionListener<ResolvedIndices> listener) {
         if (request instanceof IndicesAliasesRequest) {
-            ResolvedIndices.Builder resolvedIndicesBuilder = new ResolvedIndices.Builder();
             IndicesAliasesRequest indicesAliasesRequest = (IndicesAliasesRequest) request;
+            GroupedActionListener<ResolvedIndices> groupedListener = new GroupedActionListener<>(listener.map(allResolvedIndices -> {
+                ResolvedIndices.Builder resolvedIndicesBuilder = new ResolvedIndices.Builder();
+                for (ResolvedIndices resolvedIndices : allResolvedIndices) {
+                    resolvedIndicesBuilder.add(resolvedIndices);
+                }
+                return resolvedIndicesBuilder.build();
+            }), indicesAliasesRequest.getAliasActions().size());
             for (IndicesRequest indicesRequest : indicesAliasesRequest.getAliasActions()) {
-                final ResolvedIndices resolved = resolveIndicesAndAliases(indicesRequest, metadata, authorizedIndices);
-                resolvedIndicesBuilder.addLocal(resolved.getLocal());
-                resolvedIndicesBuilder.addRemote(resolved.getRemote());
+                resolveIndicesAndAliases(indicesRequest, metadata, authorizedIndicesSupplier, groupedListener);
+            }
+        } else {
+            // if for some reason we are missing an action... just for safety we'll reject
+            if (request instanceof IndicesRequest == false) {
+                listener.onFailure(new IllegalStateException("Request [" + request + "] is not an Indices request, but should be."));
+            } else {
+                resolveIndicesAndAliases((IndicesRequest) request, metadata, authorizedIndicesSupplier, listener);
             }
-            return resolvedIndicesBuilder.build();
-        }
-
-        // if for some reason we are missing an action... just for safety we'll reject
-        if (request instanceof IndicesRequest == false) {
-            throw new IllegalStateException("Request [" + request + "] is not an Indices request, but should be.");
         }
-        return resolveIndicesAndAliases((IndicesRequest) request, metadata, authorizedIndices);
     }
 
+    ResolvedIndices resolve(TransportRequest request, Metadata metadata, List<String> authorizedIndices) {
+        PlainActionFuture<ResolvedIndices> future = PlainActionFuture.newFuture();
+        // if the supplier is not async, the method is not async, hence get is non-blocking
+        resolve(request, metadata, listener -> listener.onResponse(authorizedIndices), future);
+        return FutureUtils.get(future, 0, TimeUnit.MILLISECONDS);
```

Comment:
Sort of a hack:
If the authorized supplier is not in fact async (`listener -> listener.onResponse(authorizedIndices)`) we know (assume) that the `resolve` invocation is going to be sync. Hence, we can call `FutureUtils.get`. But this calls the timeout variant, because the get without timeout variant, asserts that the calling thread supports blocking (which it usually doesn't)